### PR TITLE
[Test] contract staking test contract

### PR DESCRIPTION
**Pull request by EdHastingsCasperAssociation**
This PR adds a test contract for contract staking to prevent regression.

### CODE REVIEW

File: storage/src/system/protocol_upgrade.rs [hunk 1]
```diff
@@ -1132,6 +1133,38 @@ where
...
+        let tc = &mut self.tracking_copy;
+        let existing_keys = match tc.get_keys(&KeyTag::Unbond) {
+            Ok(keys) => keys,
+            Err(err) => return Err(ProtocolUpgradeError::TrackingCopy(err)),
+        };
+        for key in existing_keys {
+            if let Some(StoredValue::Unbonding(unbonding_purses)) =
+                tc.get(&key).map_err(Into::<ProtocolUpgradeError>::into)?
+            {
+                // prune away the original record, we don't need it anymore
```
Comment (darthsiroftardis) [inline storage/src/system/protocol_upgrade.rs:1148]: Should we have an if condition to only migrate unbonding purses based on the unbonding delay, i.e what we did for redelegate, only migrate the purses that are yet to be paid

**Review by darthsiroftardis — APPROVED** [review]

File: storage/src/system/protocol_upgrade.rs (hunk 1 above)
Comment (EdHastingsCasperAssociation) [reply storage/src/system/protocol_upgrade.rs:1148]: Maybe. However, the original logic was a stored value of a vec of unbonding purses which were being popped off for a pseudo prune so there shouldn't be any such records.

On that note tho, an optimization to not write the record if unbond.eras().len() == 0 would certainly be feasible.

File: storage/src/system/runtime_native.rs [hunk 2]
```diff
@@ -479,6 +479,11 @@ where
         &self.runtime_footprint
     }
 
+    /// Returns the addressable entity being used by this instance.
```
Comment (EdHastingsCasperAssociation) [inline storage/src/system/runtime_native.rs:482]: Just noticed this docu comment is wrong. Should be:

"Returns a mutable reference to the runtime footprint used by this instance."

**Comment by devendran-m** [comment]
Related issue #4929

**Pull request merged**